Composer/plugins/lib/bot-deploy/src/utils/crossTrainUtil.ts

```diff
@@ -0,0 +1,215 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/**
+ * luUtil.ts is a single place use lu-parser handle lu file operation.
+ * it's designed have no state, input text file, output text file.
+ * for more usage detail, please check client/__tests__/utils/luUtil.test.ts
+ */
+import keys from 'lodash/keys';
+import { LuFile, DialogInfo, IIntentTrigger, FieldNames, SDKKinds, FileInfo } from '@bfc/shared';
+import { luIndexer } from '@bfc/indexers';
+
+import { getBaseName, getExtension } from './fileUtil';
+import { VisitorFunc, JsonWalk } from './jsonWalk';
+
+export function getReferredLuFiles(luFiles: LuFile[], dialogs: DialogInfo[], checkContent = true) {
+  return luFiles.filter((file) => {
+    const idWithOutLocale = getBaseName(file.id);
+    return dialogs.some(
+      (dialog) => dialog.luFile === idWithOutLocale && ((checkContent && !!file.content) || !checkContent)
+    );
+  });
+}
+
+function ExtractAllBeginDialogs(value: any): string[] {
+  const dialogs: string[] = [];
+
+  const visitor: VisitorFunc = (path: string, value: any): boolean => {
+    if (value?.$kind === SDKKinds.BeginDialog && value?.dialog) {
+      dialogs.push(value.dialog);
+      return true;
+    }
+    return false;
+  };
+
+  JsonWalk('$', value, visitor);
+
+  return dialogs;
+}
+
+// find out all properties from given dialog
+function ExtractIntentTriggers(value: any): IIntentTrigger[] {
+  const intentTriggers: IIntentTrigger[] = [];
+  const triggers = value?.[FieldNames.Events];
+
+  if (triggers && triggers.length) {
+    for (const trigger of triggers) {
+      const dialogs = ExtractAllBeginDialogs(trigger);
+
+      if (trigger.$kind === SDKKinds.OnIntent && trigger.intent) {
+        intentTriggers.push({ intent: trigger.intent, dialogs });
+      } else if (trigger.$kind !== SDKKinds.OnIntent && dialogs.length) {
+        const emptyIntent = intentTriggers.find((e) => e.intent === '');
+        if (emptyIntent) {
+          //remove the duplication dialogs
+          const all = new Set<string>([...emptyIntent.dialogs, ...dialogs]);
+          emptyIntent.dialogs = Array.from(all);
+        } else {
+          intentTriggers.push({ intent: '', dialogs });
+        }
+      }
+    }
+  }
```

Comment:
Duplications:
![image](https://user-images.githubusercontent.com/8528761/90368690-0bc26d00-e09d-11ea-81cc-c6e0390f86e8.png)

@lei9444 is it allowed to import `indexer` in `bot-deploy`?

Reply:
The same problem with jsonWalk.ts